File: 6_zbieranie_danych_REST_API/from_c++/rest_api.cpp
#include "crow.h"
#include "nlohmann/json.hpp"
#include <fstream>
#include <vector>
#include <algorithm>

using json = nlohmann::json;
const std::string DATA_FILE = "users.json";

// Load users from file
json load_users() {
    std::ifstream file(DATA_FILE);
    if (!file.is_open()) return json::array();
    json users;
    file >> users;
    return users;
}

// Save users to file
void save_users(const json& users) {
    std::ofstream file(DATA_FILE);
    file << users.dump(4);
}

int main() {
    crow::SimpleApp app;

    // GET /users – list of users
    CROW_ROUTE(app, "/users").methods(crow::HTTPMethod::GET)
        ([]() {
        json users = load_users();
        return crow::response{ users.dump() };
            });

    // GET /users/<id> – single user
    CROW_ROUTE(app, "/users/<int>").methods(crow::HTTPMethod::GET)
        ([](int user_id) {
        json users = load_users();
        auto it = std::find_if(users.begin(), users.end(), [&](const json& u) { return u["id"] == user_id; });
        if (it != users.end()) {
            return crow::response{ it->dump() };
        }
        return crow::response(404, R"({"error":"User not found"})");
            });

    // POST /users – adding user
    CROW_ROUTE(app, "/users").methods(crow::HTTPMethod::POST)
        ([](const crow::request& req) {
        auto data = json::parse(req.body, nullptr, false);
        if (!data.is_object() || !data.contains("name") || !data.contains("email"))
            return crow::response(400, R"({"error":"Missing name or email"})");

        json users = load_users();
        int new_id = 0;
        for (auto& u : users) new_id = std::max(new_id, u["id"].get<int>());
        new_id++;

        json new_user = { {"id", new_id}, {"name", data["name"]}, {"email", data["email"]} };
        users.push_back(new_user);
        save_users(users);

        return crow::response(201, new_user.dump());
            });

    // PUT /users/<id> – update user
    CROW_ROUTE(app, "/users/<int>").methods(crow::HTTPMethod::PUT)
        ([](const crow::request& req, int user_id) {
        auto data = json::parse(req.body, nullptr, false);
        if (!data.is_object())
            return crow::response(400, R"({"error":"Invalid JSON"})");

        json users = load_users();
        for (auto& u : users) {
            if (u["id"] == user_id) {
                if (data.contains("name")) u["name"] = data["name"];
                if (data.contains("email")) u["email"] = data["email"];
                save_users(users);
                return crow::response{ u.dump() };
            }
        }
        return crow::response(404, R"({"error":"User not found"})");
            });

    // DELETE /users/<id> – delete user
    CROW_ROUTE(app, "/users/<int>").methods(crow::HTTPMethod::Delete)
        ([](int user_id) {
        json users = load_users();
        size_t before = users.size();
        users.erase(std::remove_if(users.begin(), users.end(),
            [&](const json& u) { return u["id"] == user_id; }),
            users.end());
        if (users.size() == before)
            return crow::response(404, R"({"error":"User not found"})");

        save_users(users);
        return crow::response(R"({"message":"User deleted"})");
            });

    app.port(5000).multithreaded().run();
}

File: 6_zbieranie_danych_REST_API/from_python/rest_api.py
from flask import Flask, request, jsonify
import json
import os

app = Flask(__name__)
DATA_FILE = "users.json"


# Help functions
def load_users():
    if not os.path.exists(DATA_FILE):
        return []
    with open(DATA_FILE, "r") as f:
        return json.load(f)


def save_users(users):
    with open(DATA_FILE, "w") as f:
        json.dump(users, f, indent=4)


# Endpoint GET - user list
@app.route("/users", methods=["GET"])
def get_users():
    users = load_users()
    return jsonify(users), 200


# Endpoint GET – single user
@app.route("/users/<int:user_id>", methods=["GET"])
def get_user(user_id):
    users = load_users()
    user = next((u for u in users if u["id"] == user_id), None)
    if user:
        return jsonify(user), 200
    return jsonify({"error": "User not found"}), 404


# Endpoint POST – add users
@app.route("/users", methods=["POST"])
def add_user():
    data = request.get_json()
    if not data or "name" not in data or "email" not in data:
        return jsonify({"error": "Missing name or email"}), 400

    users = load_users()
    new_id = max([u["id"] for u in users], default=0) + 1
    new_user = {"id": new_id, "name": data["name"], "email": data["email"]}
    users.append(new_user)
    save_users(users)

    return jsonify(new_user), 201


# Endpoint PUT – edit user
@app.route("/users/<int:user_id>", methods=["PUT"])
def update_user(user_id):
    data = request.get_json()
    users = load_users()
    for u in users:
        if u["id"] == user_id:
            u["name"] = data.get("name", u["name"])
            u["email"] = data.get("email", u["email"])
            save_users(users)
            return jsonify(u), 200

    return jsonify({"error": "User not found"}), 404


# Endpoint DELETE – delete user
@app.route("/users/<int:user_id>", methods=["DELETE"])
def delete_user(user_id):
    users = load_users()
    new_users = [u for u in users if u["id"] != user_id]
    if len(new_users) == len(users):
        return jsonify({"error": "User not found"}), 404

    save_users(new_users)
    return jsonify({"message": "User deleted"}), 200


if __name__ == "__main__":
    app.run(debug=True)
